### DISCUSSION TITLE
Any tips for avoiding singular matrix errors

### DISCUSSION THREAD

**Discussion by laat0003** [pypest/pyemu#501] [Q&A] [answered]
Hi,

I'm trying to write a PEST style uncertainty file through PstFrom.build_prior using uncfile as the format but linalg doesn't want to play nice. It keeps throwing singular matrix errors at me. 

The errors appear to be related to grid-scale multiplier parameterization of boundary conditions for which I have a geostruct assigned. 

I was hoping you could advise how I might avoid these errors or alternatively suggest a different way to get an uncfile out of Pyemu.

Thanks in advance for your time,
Tariq

**Comment by jtwhite79** [comment, marked as answer]
Usually (usually...) that error/warning is because there are parameters that have zero separation distance (they are colocated in space or time) which yields a singular cov matrix from the geostruct.   Maybe you have colocated boundary parameters?  If this is the case, I think you can tie these parameters together before forming the cov matrix (if tying them together is acceptable for you?)...

**Reply by laat0003** [reply]
Thanks for that @jtwhite79.

I'll take a closer look at the problematic pars. I assumed that if I had no issues building a prior ensemble with PstFrom.draw then I shouldn't have issues with the uncfile creation. I tried dropping the geostruct from the grid-scale pars and it powered through like a champ but the cov file is 8GB so perhaps I need to rethink my boundary parameterization strategy. :)

**Comment by laat0003** [comment]
You were 100% correct @jtwhite79. The issue was colocated parameters along a perimeter boundary in a multi-layer model. Tying them wasn't an option in this instance so relied on use_rows instead to assign layer specific boundary parameters. Many thanks for pointing me in the right direction.
